Using from now on q as exit key

maccxs opened the pull request:
* made this work for all levels
* moved all the leveltimeout delayed endlevel stuff into the end level controller of sandbox
* each level MUST have EndLevel() implemented, this led to errors before, since in the
end level base EndLevel() called the end screen of the sandbox
* end level base should now  yield an error message if this method is called in the base class
* this may also be linked to #108

Commit:
Using from now on q as exit key

* made this work for all levels
* moved all the leveltimeout delayed endlevel stuff into the end level controller of sandbox
* each level MUST have EndLevel() implemented, this led to errors before, since in the
end level base EndLevel() called the end screen of the sandbox
* end level base should now  yield an error message if this method is called in the base class

Commit:
q key now also works in sandbox mode

holke commented on YouVsVirus/Assets/Scenes/Levelgethome/EndLevelControllerLevelgethome.cs:
```diff
@@ -9,6 +10,12 @@
+    {   
+        CanvasFail.SetActive(true);
+        failText.text = "You pressed the exit key.";
+    }
```
What if we end the level in a different way than pressing q?
I.e. when the player enters the house or is infected.
I think from the logic of the elc and the naming of this function, all those cases should call `EndLevel()`
and the function should then act appropiately.
So maybe this particular action should be outsourced to a function `ExitKeyPressed()` which is called by `EndLevel()` if the key was pressed and if the level was ended differently something else happens.

maccxs replied:
> 
> 
> > I tried with pausing the game, but then if I press "Retry" the game is still paused. Any ideas how to fix that?
> 
> The retry button should then call Unpause when pressed.

Doing this in Retry() is easy, since this function is in the base class. We also have to do it for Continue() in every level...

maccxs replied:
Done

holke commented on YouVsVirus/Assets/Scenes/leveldisco/EndLevelControllerLeveldisco.cs:
```diff
@@ -10,7 +10,11 @@
+    {
+        CanvasFail.SetActive(true);
+        failText.text = "You pressed the exit key.";
+    }
```
The same  comment as for the gethome level

maccxs replied:
Done

holke commented on YouVsVirus/Assets/Scenes/levelsupermarket/EndLevelControllerLevelsupermarket.cs:
```diff
@@ -10,6 +11,17 @@
+        PressSpaceCanv.SetActive(false);
+        CanvasFail.SetActive(true);
+        failText.text = "You pressed the exit key.";
+    }
```
same comment

maccxs replied:
Done

holke commented on YouVsVirus/Assets/Scripts/Components/EndLevelControllerBase.cs:
```diff
@@ -60,11 +39,7 @@ public void Awake()
+
 
     /// <summary>
     /// Triggers the end of the level.
```
This comment should now state that this function will cause an error and that 
a derived class must implement it.

maccxs replied:
Removed endlevel from base. Now only exits in sandbox

holke commented on YouVsVirus/Assets/Scripts/Components/EndLevelControllerBase.cs:
```diff
@@ -73,19 +48,9 @@ public void Start()
     public virtual void EndLevel()
```
I think a different way would be
```suggestion
    public abstract void EndLevel();
```
and remove the function footer.
This way we will get a compiler error when a derived class does not implement EndLevel.

maccxs replied:
Removed EndLevel(), but good idea of course

holke commented:
At least in the supermarket level i can continue playing the game and even finish it after i pressed q.
The exit canvas pops up, but everything else keeps on running in the background.
I did not test it for the other level but guess it is the same there.

Maybe resolve this by pausing the game when q is pressed?

Commit:
Basic integration of master into leveldemo

Commit:
Adapted end conditions and player house

Commit:
merge part 1

maccxs commented:
I tried with pausing the game, but then if I press "Retry" the game is still paused. Any ideas how to fix that?

Commit:
Merge branch 'leveldemo2' into margrit/exitkey

holke reviewed (commented):
> 
> 
> I tried with pausing the game, but then if I press "Retry" the game is still paused. Any ideas how to fix that?

The retry button should then call Unpause when pressed.

Commit:
Tweaks for end game

* GAme is paused and unpaused if q or c is pressed
* C continues the game with success
* EndLevel now only lives in sandbox
* New function "ExitKey" implemented
* Moved these ending conditions into the base class
* Text component is handled automatically

Commit:
Added end screen prototype

Commit:
Forgot to save scene

holke commented:
In sandbox mode when i press 'c' or 'q' twice then it ends. I restart and it ends after 5  seconds again.
Possibly because the input is cached somewhere and this is then read in the next round as a key press.
Thus, we should find out how to delete this internal cache and do so before a level starts.

holke commented on YouVsVirus/Assets/Scripts/Components/EndLevelControllerBase.cs:
```diff
@@ -179,70 +147,97 @@ protected void EndGamePlayerExposed()
+    protected virtual bool LevelDependentEndGameConditionFulfilled()
+    {
+        return true;
+    }
```
```suggestion
    }
    
```

holke replied:
Just adding a newline here.

holke committed:
Update YouVsVirus/Assets/Scripts/Components/EndLevelControllerBase.cs

holke committed:
Endlevelcontroller can end the level only once.

It could have happened that the elc ended the level
because Ãa button was pressed and in the same frame end
it because the playre reached the goal.

We introduced safety measures to prevent this.
A variable levelhasfinished now gets set to true
as soon as the level finishes and the elc does not
check any condition if this var is true.

holke committed:
Make test mode optional for endlevelcontroller

This way, we do not need to remove any code when
we want to switch off test mode.
We just uncheck the test mode box in the scene view
for the endlevel controller.
Also, this allows us to have test mode more some levels
and not for others if we want to.

holke committed:
Merge remote-tracking branch 'origin/margrit/exitkey' into margrit/exitkey

holke commented:
I just made two updates.

1. It was possible that the ELC triggered the EndGamePlayerExposed and EndGamePlayerAtHome simultaneously. I deactivated this behaviour. Now once one of the ending routines is called, no other can be called. 
This also fixed the issue in the sandbox mode described in my comment above.

2. The ELC now has a public variable testMode, which we can set in the scene view. If true, then pressing c will finish the level. If false it does not.
This way, we wont have to delete any code to deactivate test mode. We just need to uncheck the box for each endlevelcontroller in the level scene view of the unity editor.

holke commented:
i was too quick there. The problem still exists.

holke commented:
Ok, i know where this comes from.
The level gets ended by pressing ´c´ or ´q´ and when we then restart quickly enought, somehow, the DelayedEndLevel routine is running and terminates the new level after a few seconds.

This is particularly weird since the whole elc should be a new instance when the sandbox is restarted...
Also i tried to stop the coroutines when exiting but did not succeed.

holke commented:
I am using this as a rubber duck now...

i am getting closer to the cause of this. 
Somehow, when we restart the sandbox, the elc sometimes registers 0 active infections.
And this then triggers the CheckEndCondition and finishes the level.

holke commented:
Ok, i think i have it.
It is unrelated to pressing the buttons.

Sometimes the `Start()` Routine of `HumanBase` is called before the `Start()` routine of `HumanInstants`.
This is due to the fact, that the order in which the `Start()` routines are called is undefined.

When an NPC starts as infected, this is set by `SetInitialCondition` in `HumanInstants::CreateHumans` which is called during `Start`.

This initial infected NPC is only passed on to the EndLevelController in the NPCs `Start()` routine at
```
SetCondition(_initialCondition);
```

but, if the NPCs `Start()` routine was already executed before its `InitialCondition` was set by the `HumanBase` start, then this Info does not get passed on to the EndLevelController.
And thus, the ELC thinks that there are 0 infected NPCs and hence wants to finish the game.

This also causes the ELC to think that there are less infected humans then actually are.

holke commented:
I think this is unrelated to the q key and opened a new issue.

holke commented:
I think we can merge this as soon as leveldemo is reviewed

holke committed:
sandbox elc: replace endConditionMet with levelHasFinished

holke committed:
Stop coroutines in sandbox elc when level ends

holke commented:
If i get infected in demo level and then retry the game is paused

holke committed:
Update PauseGame with gameIsPaused state

holke commented:
> 
> 
> If i get infected in demo level and then retry the game is paused

Solved by improving the PauseGame class.

Commit:
Jeff is now Sandra and the infection fine-tuned

maccxs commented:
> 
> 
> > If i get infected in demo level and then retry the game is paused
> 
> Solved by improving the PauseGame class.

Did you already commit this? I have the same problem in the disco level

Commit:
More finetune of spread

Commit:
Adapted start screen design to a cleaner look
* similar script size in all levels
* smaller panel size

Commit:
Canvas darker in every level´, enough?

Commit:
Introduced a bool that makes sure that the sandbox is not ended early

Commit:
Made demolevel a little harder
* more infected
* also smileys without sign have chance of propaganda infection
* removed some empty space
* removed no longer requuired string comparison for leveldisco

Commit:
Fixes Levelcollectmasks

* during merge the mask text counter got lost
* cummulative sprite update was lost due to changes in endgame controller, now fixed
* masks needed was set to 1 now again to 8
* cannot unpause the game anymore :-( anywhere

Commit:
game is paused was never set to true

maccxs commented:
> 
> 
> > If i get infected in demo level and then retry the game is paused
> 
> Solved by improving the PauseGame class.

This really puzzled me since it got worse after your commit. However, gamePaused was not set to true in your code. Fixed in 
70c33e5

Commit:
Forgotten scene

Commit:
Made demolevel even harder by placing barriers differently

Commit:
Adapted script size of counter text in levelsupermarket/collectmasks

holke commented on YouVsVirus/Assets/Scenes/leveldisco/EndLevelControllerLeveldisco.cs:
```diff
@@ -49,5 +49,17 @@ public void PlayerExitSickWithoutFriend()
     }
 
+    /// <summary>
+    /// In infection status is not shown all NCPs do no sprite update within the game
```
```suggestion
    /// If infection status is not shown all NCPs do no sprite update within the game
```
?

holke commented on YouVsVirus/Assets/Scripts/Infection/ProbabilityBasedInfection.cs:
```diff
@@ -34,10 +34,7 @@ public override void Update()
-                // due to the much shorter time scale
-                if(LevelSettings.GetActiveSceneName() != "YouVsVirus_Leveldisco")
-                    UpdateCondition();
+                UpdateCondition();
```
Can you explain why you changed this?
Does not seem to have to do anything with the Q-key or the demo level...

maccxs replied:
Should have done it separately. Was superfluous since we handle this in the level settings now.

holke commented on YouVsVirus/Assets/Scenes/leveldemo/StartScreenLeveldemo.unity:
```diff
@@ -0,0 +1,1703 @@
+      m_Calls: []
+  m_text: 'You come home from your daily walk and you cannot believe your eyes: There
+    is some kind of demonstration going on and your house is right in the middle.
+    Surly it won''t hurt to try to pass these strange people, will it? '
```
```suggestion
    Surely it won''t hurt to try to pass these strange people, will it? '
```

holke reviewed (approved):
Overall this plays nice and everything seems to work.

Just minor edits.

After we are through with this we should however discuss our branching and reviewing workflow and the public NPC number stuff.

holke committed:
Typo

Commit:
Confetti instead of balloons

maccxs commented:
> 
> 
> Damn, thats embarrassing for me :(

Dont't worry about it.

> ter we are through with this we should however discuss our branching and reviewing workflow and the public NPC number stuff.

Yes, sorry that this was so chaotic. We should definitely fix some guidlines which also take into account that merging stuff in unity is a pain.

maccxs committed:
Update YouVsVirus/Assets/Scenes/leveldisco/EndLevelControllerLeveldisco.cs

Co-authored-by: holke <johannes.holke@dlr.de>

Commit:
Maybe a little harder?

Commit:
Added stay safe message

maccxs committed:
Merge pull request #129 from youvsvirus/margrit/discolevel-finetune

Margrit/discolevel finetune

Commit:
Merge branch 'margrit/exitkey' into margrit/design

Commit:
Made Jeff to Sandra

maccxs committed:
Merge branch 'margrit/exitkey' into margrit/design

maccxs committed:
Merge pull request #131 from youvsvirus/margrit/design

Margrit/design
